atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/BatchedDbReadTable.java

```diff
@@ -0,0 +1,175 @@
+    public ClosableIterator<AgnosticLightResultRow> getLatestCellsInternal(Map<Cell, Long> cells,
+                                                                           boolean includeValue) {
+        Queue<Future<ClosableIterator<AgnosticLightResultRow>>> futures = Queues.newArrayDeque();
+        for (final List<Entry<Cell, Long>> batch : Iterables.partition(cells.entrySet(), getBatchSize())) {
+            futures.add(submit(exec, queryFactory.getLatestCellsQuery(batch, includeValue)));
+        }
+        return new LazyClosableIterator<AgnosticLightResultRow>(futures);
+    }
+
+    @Override
+    public ClosableIterator<AgnosticLightResultRow> getAllCellsInternal(Iterable<Cell> cells,
+                                                                        long ts,
+                                                                        boolean includeValue) {
+        Queue<Future<ClosableIterator<AgnosticLightResultRow>>> futures = Queues.newArrayDeque();
+        for (final List<Cell> batch : Iterables.partition(cells, getBatchSize())) {
+            futures.add(submit(exec, queryFactory.getAllCellsQuery(batch, ts, includeValue)));
+        }
+        return new LazyClosableIterator<AgnosticLightResultRow>(futures);
+    }
+
+    @Override
+    public ClosableIterator<AgnosticLightResultRow> getAllCellsInternal(Map<Cell, Long> cells,
+                                                                        boolean includeValue) {
+        Queue<Future<ClosableIterator<AgnosticLightResultRow>>> futures = Queues.newArrayDeque();
+        for (final List<Entry<Cell, Long>> batch : Iterables.partition(cells.entrySet(), getBatchSize())) {
+            futures.add(submit(exec, queryFactory.getAllCellsQuery(batch, includeValue)));
+        }
+        return new LazyClosableIterator<AgnosticLightResultRow>(futures);
+    }
+
+    @Override
+    protected ClosableIterator<AgnosticLightResultRow> run(FullQuery query) {
+        final PalantirSqlConnection freshConn = conns.getFresh();
+        try {
+            final AgnosticLightResultSet results = freshConn.selectLightResultSetUnregisteredQuery(
+                    query.getQuery(), query.getArgs());
+            return ClosableIterators.wrap(results.iterator(), new Closeable() {
+                @Override
+                public void close() {
+                    results.close();
+                    closeConnection(freshConn.getUnderlyingConnection());
+                }
+            });
+        } catch (Exception e) {
+            closeConnection(freshConn.getUnderlyingConnection());
+            throw Throwables.rewrapAndThrowUncheckedException(e);
+        }
+    }
+
+    private void closeConnection(Connection conn) {
+        try {
+            conn.close();
+        } catch (SQLException e) {
+            log.error("Failed to close db connection performing reads.", e);
+        }
+    }
+
+    private int getBatchSize() {
+        return config.postgresQueryPoolSize();
```

Comment:
I don't think this is correct, this should be the batch size from config, not pool size